File: week_2/code/day_8/f_design_twitter.py
import heapq
from collections import defaultdict

class Twitter:
    def __init__(self):
        # Global counter to act as a timestamp (lower means older)
        self.time = 0
        self.tweets = defaultdict(list) # userId -> list of [timestamp, tweetId]
        self.followees = defaultdict(set) # userId -> set of followeeIds

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweets[userId].append([self.time, tweetId])
        self.time += 1

    def getNewsFeed(self, userId: int) -> list[int]:
        # A user automatically follows themselves for the news feed
        users_to_check = self.followees[userId].copy()
        users_to_check.add(userId)
        
        # Max-heap to get the 10 most recent tweets
        # (Stored as max-heap using -timestamp)
        max_heap = []
        for uId in users_to_check:
            # Add up to the 10 most recent tweets of each followed user
            user_tweets = self.tweets[uId][-10:]
            for timestamp, tId in user_tweets:
                max_heap.append([-timestamp, tId])
        
        heapq.heapify(max_heap)
        
        # Extract at most 10 most recent tweets
        feed = []
        while max_heap and len(feed) < 10:
            feed.append(heapq.heappop(max_heap)[1])
            
        return feed

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId != followeeId:
            self.followees[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.followees[followerId]:
            self.followees[followerId].remove(followeeId)

# Example Usage
if __name__ == "__main__":
    twitter = Twitter()
    twitter.postTweet(1, 5)  # User 1 posts tweet 5
    print(twitter.getNewsFeed(1))  # return [5]
    twitter.follow(1, 2)     # User 1 follows User 2
    twitter.postTweet(2, 6)  # User 2 posts tweet 6
    print(twitter.getNewsFeed(1))  # return [6, 5]
    twitter.unfollow(1, 2)   # User 1 unfollows User 2
    print(twitter.getNewsFeed(1))  # return [5]

File: week_2/code/day_8/f_design_twitter.js
const MinHeap = require('./min_heap');

var Twitter = function () {
    this.time = 0;
    this.tweets = new Map(); // userId -> Array of { timestamp, tweetId }
    this.followees = new Map(); // userId -> Set of followeeIds
};

/** 
 * @param {number} userId 
 * @param {number} tweetId
 * @return {void}
 */
Twitter.prototype.postTweet = function (userId, tweetId) {
    if (!this.tweets.has(userId)) {
        this.tweets.set(userId, []);
    }
    this.tweets.get(userId).push({ timestamp: this.time, tweetId: tweetId });
    this.time++;
};

/** 
 * @param {number} userId
 * @return {number[]}
 */
Twitter.prototype.getNewsFeed = function (userId) {
    const usersToCheck = new Set(this.followees.get(userId) || []);
    usersToCheck.add(userId);

    // We want a Max-Heap based on timestamp (largest timestamp first)
    // So the comparator should return a value > 0 if b > a => b.timestamp - a.timestamp
    const maxHeap = new MinHeap((a, b) => b.timestamp - a.timestamp);

    for (let uId of usersToCheck) {
        if (this.tweets.has(uId)) {
            const userTweets = this.tweets.get(uId);
            // Get at most 10 recent
            const recent = userTweets.slice(-10);
            for (let t of recent) {
                maxHeap.push(t);
            }
        }
    }

    const feed = [];
    while (maxHeap.size() > 0 && feed.length < 10) {
        feed.push(maxHeap.pop().tweetId);
    }

    return feed;
};

/** 
 * @param {number} followerId 
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.follow = function (followerId, followeeId) {
    if (followerId === followeeId) return;
    if (!this.followees.has(followerId)) {
        this.followees.set(followerId, new Set());
    }
    this.followees.get(followerId).add(followeeId);
};

/** 
 * @param {number} followerId 
 * @param {number} followeeId
 * @return {void}
 */
Twitter.prototype.unfollow = function (followerId, followeeId) {
    if (this.followees.has(followerId)) {
        this.followees.get(followerId).delete(followeeId);
    }
};

// Example Usage
const twitter = new Twitter();
twitter.postTweet(1, 5); // User 1 posts tweet 5
console.log(twitter.getNewsFeed(1)); // return [5]
twitter.follow(1, 2);    // User 1 follows User 2
twitter.postTweet(2, 6); // User 2 posts tweet 6
console.log(twitter.getNewsFeed(1)); // return [6, 5]
twitter.unfollow(1, 2);  // User 1 unfollows User 2
console.log(twitter.getNewsFeed(1)); // return [5]
